File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/java/io/trino/tpcds/row/CallCenterRow.java
package io.trino.tpcds.row;

import io.trino.tpcds.type.Address;
import io.trino.tpcds.type.Decimal;
import org.apache.kyuubi.spark.connector.tpcds.row.KyuubiTPCDSTableRowWithNulls;

import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_ADDRESS;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_CALL_CENTER_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_CALL_CENTER_SK;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_CITY;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_CLASS;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_CLOSED_DATE_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_COMPANY;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_COMPANY_NAME;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_COUNTRY;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_DIVISION;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_DIVISION_NAME;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_EMPLOYEES;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_GMT_OFFSET;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_HOURS;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_MANAGER;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_MARKET_CLASS;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_MARKET_DESC;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_MARKET_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_MARKET_MANAGER;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_NAME;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_OPEN_DATE_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_REC_END_DATE_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_REC_START_DATE_ID;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_SQ_FT;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_STATE;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_STREET_NAME;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_STREET_NUMBER;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_STREET_TYPE;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_SUITE_NUMBER;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_TAX_PERCENTAGE;
import static io.trino.tpcds.generator.CallCenterGeneratorColumn.CC_ZIP;

public class CallCenterRow extends KyuubiTPCDSTableRowWithNulls {
  private final long ccCallCenterSk;
  private final String ccCallCenterId;
  private final long ccRecStartDateId;
  private final long ccRecEndDateId;
  private final long ccClosedDateId;
  private final long ccOpenDateId;
  private final String ccName;
  private final String ccClass;
  private final int ccEmployees;
  private final int ccSqFt;
  private final String ccHours;
  private final String ccManager;
  private final int ccMarketId;
  private final String ccMarketClass;
  private final String ccMarketDesc;
  private final String ccMarketManager;
  private final int ccDivisionId;
  private final String ccDivisionName;
  private final int ccCompany;
  private final String ccCompanyName;
  private final Address ccAddress;
  private final Decimal ccTaxPercentage;

  private CallCenterRow(long ccCallCenterSk,
      String ccCallCenterId,
      long ccRecStartDateId,
      long ccRecEndDateId,
      long ccClosedDateId,
      long ccOpenDateId,
      String ccName,
      String ccClass,
      int ccEmployees,
      int ccSqFt,
      String ccHours,
      String ccManager,
      int ccMarketId,
      String ccMarketClass,
      String ccMarketDesc,
      String ccMarketManager,
      int ccDivisionId,
      String ccDivisionName,
      int ccCompany,
      String ccCompanyName,
      Address ccAddress,
      Decimal ccTaxPercentage,
      long nullBitMap)
  {
    super(nullBitMap, CC_CALL_CENTER_SK);
    this.ccCallCenterSk = ccCallCenterSk;
    this.ccCallCenterId = ccCallCenterId;
    this.ccRecStartDateId = ccRecStartDateId;
    this.ccRecEndDateId = ccRecEndDateId;
    this.ccClosedDateId = ccClosedDateId;
    this.ccOpenDateId = ccOpenDateId;
    this.ccName = ccName;
    this.ccClass = ccClass;
    this.ccEmployees = ccEmployees;
    this.ccSqFt = ccSqFt;
    this.ccHours = ccHours;
    this.ccManager = ccManager;
    this.ccMarketId = ccMarketId;
    this.ccMarketClass = ccMarketClass;
    this.ccMarketDesc = ccMarketDesc;
    this.ccMarketManager = ccMarketManager;
    this.ccDivisionId = ccDivisionId;
    this.ccDivisionName = ccDivisionName;
    this.ccCompany = ccCompany;
    this.ccCompanyName = ccCompanyName;
    this.ccAddress = ccAddress;
    this.ccTaxPercentage = ccTaxPercentage;
  }

  public long getCcCallCenterSk()
  {
    return ccCallCenterSk;
  }

  public String getCcCallCenterId()
  {
    return ccCallCenterId;
  }

  public long getCcRecStartDateId()
  {
    return ccRecStartDateId;
  }

  public long getCcRecEndDateId()
  {
    return ccRecEndDateId;
  }

  public long getCcClosedDateId()
  {
    return ccClosedDateId;
  }

  public long getCcOpenDateId()
  {
    return ccOpenDateId;
  }

  public String getCcName()
  {
    return ccName;
  }

  public String getCcClass()
  {
    return ccClass;
  }

  public int getCcEmployees()
  {
    return ccEmployees;
  }

  public int getCcSqFt()
  {
    return ccSqFt;
  }

  public String getCcHours()
  {
    return ccHours;
  }

  public String getCcManager()
  {
    return ccManager;
  }

  public int getCcMarketId()
  {
    return ccMarketId;
  }

  public String getCcMarketClass()
  {
    return ccMarketClass;
  }

  public String getCcMarketDesc()
  {
    return ccMarketDesc;
  }

  public String getCcMarketManager()
  {
    return ccMarketManager;
  }

  public int getCcDivisionId()
  {
    return ccDivisionId;
  }

  public String getCcDivisionName()
  {
    return ccDivisionName;
  }

  public int getCcCompany()
  {
    return ccCompany;
  }

  public String getCcCompanyName()
  {
    return ccCompanyName;
  }

  public Address getCcAddress()
  {
    return ccAddress;
  }

  public Decimal getCcTaxPercentage()
  {
    return ccTaxPercentage;
  }


  public static class Builder
  {
    private long ccCallCenterSk;
    private String ccCallCenterId;
    private long ccRecStartDateId;
    private long ccRecEndDateId;
    private long ccClosedDateId;
    private long ccOpenDateId;
    private String ccName;
    private String ccClass;
    private int ccEmployees;
    private int ccSqFt;
    private String ccHours;
    private String ccManager;
    private int ccMarketId;
    private String ccMarketClass;
    private String ccMarketDesc;
    private String ccMarketManager;
    private int ccDivisionId;
    private String ccDivisionName;
    private int ccCompany;
    private String ccCompanyName;
    private Address ccAddress;
    private Decimal ccTaxPercentage;
    private long nullBitMap;

    public Builder setCcCallCenterSk(long ccCallCenterSk)
    {
      this.ccCallCenterSk = ccCallCenterSk;
      return this;
    }

    public Builder setCcCallCenterId(String ccCallCenterId)
    {
      this.ccCallCenterId = ccCallCenterId;
      return this;
    }

    public Builder setCcRecStartDateId(long ccRecStartDateId)
    {
      this.ccRecStartDateId = ccRecStartDateId;
      return this;
    }

    public Builder setCcRecEndDateId(long ccRecEndDateId)
    {
      this.ccRecEndDateId = ccRecEndDateId;
      return this;
    }

    public Builder setCcClosedDateId(long ccClosedDateId)
    {
      this.ccClosedDateId = ccClosedDateId;
      return this;
    }

    public Builder setCcOpenDateId(long ccOpenDateId)
    {
      this.ccOpenDateId = ccOpenDateId;
      return this;
    }

    public Builder setCcName(String ccName)
    {
      this.ccName = ccName;
      return this;
    }

    public Builder setCcClass(String ccClass)
    {
      this.ccClass = ccClass;
      return this;
    }

    public Builder setCcEmployees(int ccEmployees)
    {
      this.ccEmployees = ccEmployees;
      return this;
    }

    public Builder setCcSqFt(int ccSqFt)
    {
      this.ccSqFt = ccSqFt;
      return this;
    }

    public Builder setCcHours(String ccHours)
    {
      this.ccHours = ccHours;
      return this;
    }

    public Builder setCcManager(String ccManager)
    {
      this.ccManager = ccManager;
      return this;
    }

    public Builder setCcMarketId(int ccMarketId)
    {
      this.ccMarketId = ccMarketId;
      return this;
    }

    public Builder setCcMarketClass(String ccMarketClass)
    {
      this.ccMarketClass = ccMarketClass;
      return this;
    }

    public Builder setCcMarketDesc(String ccMarketDesc)
    {
      this.ccMarketDesc = ccMarketDesc;
      return this;
    }

    public Builder setCcMarketManager(String ccMarketManager)
    {
      this.ccMarketManager = ccMarketManager;
      return this;
    }

    public Builder setCcDivisionId(int ccDivisionId)
    {
      this.ccDivisionId = ccDivisionId;
      return this;
    }

    public Builder setCcDivisionName(String ccDivisionName)
    {
      this.ccDivisionName = ccDivisionName;
      return this;
    }

    public Builder setCcCompany(int ccCompany)
    {
      this.ccCompany = ccCompany;
      return this;
    }

    public Builder setCcCompanyName(String ccCompanyName)
    {
      this.ccCompanyName = ccCompanyName;
      return this;
    }

    public Builder setCcAddress(Address ccAddress)
    {
      this.ccAddress = ccAddress;
      return this;
    }

    public Builder setCcTaxPercentage(Decimal ccTaxPercentage)
    {
      this.ccTaxPercentage = ccTaxPercentage;
      return this;
    }

    public CallCenterRow build()
    {
      return new CallCenterRow(ccCallCenterSk, ccCallCenterId, ccRecStartDateId, ccRecEndDateId, ccClosedDateId, ccOpenDateId, ccName, ccClass, ccEmployees, ccSqFt, ccHours, ccManager, ccMarketId, ccMarketClass, ccMarketDesc, ccMarketManager, ccDivisionId, ccDivisionName, ccCompany, ccCompanyName, ccAddress, ccTaxPercentage, nullBitMap);
    }

    public void setNullBitMap(long nullBitMap)
    {
      this.nullBitMap = nullBitMap;
    }
  }

  @Override
  public Object[] values() {
    return new Object[]{ getOrNullForKey(ccCallCenterSk, CC_CALL_CENTER_SK),
        getOrNull(ccCallCenterId, CC_CALL_CENTER_ID),
        getDateOrNullFromJulianDays(ccRecStartDateId, CC_REC_START_DATE_ID),
        getDateOrNullFromJulianDays(ccRecEndDateId, CC_REC_END_DATE_ID),
        getOrNullForKey(ccClosedDateId, CC_CLOSED_DATE_ID),
        getOrNullForKey(ccOpenDateId, CC_OPEN_DATE_ID),
        getOrNull(ccName, CC_NAME),
        getOrNull(ccClass, CC_CLASS),
        getOrNull(ccEmployees, CC_EMPLOYEES),
        getOrNull(ccSqFt, CC_SQ_FT),
        getOrNull(ccHours, CC_HOURS),
        getOrNull(ccManager, CC_MANAGER),
        getOrNull(ccMarketId, CC_MARKET_ID),
        getOrNull(ccMarketClass, CC_MARKET_CLASS),
        getOrNull(ccMarketDesc, CC_MARKET_DESC),
        getOrNull(ccMarketManager, CC_MARKET_MANAGER),
        getOrNull(ccDivisionId, CC_DIVISION),
        getOrNull(ccDivisionName, CC_DIVISION_NAME),
        getOrNull(ccCompany, CC_COMPANY),
        getOrNull(ccCompanyName, CC_COMPANY_NAME),
        getOrNull(ccAddress.getStreetNumber(), CC_STREET_NUMBER),
        getOrNull(ccAddress.getStreetName(), CC_STREET_NAME),
        getOrNull(ccAddress.getStreetType(), CC_STREET_TYPE),
        getOrNull(ccAddress.getSuiteNumber(), CC_SUITE_NUMBER),
        getOrNull(ccAddress.getCity(), CC_CITY),
        getOrNull(ccAddress.getCounty(), CC_ADDRESS),
        getOrNull(ccAddress.getState(), CC_STATE),
        getOrNull(java.lang.String.format("%05d", ccAddress.getZip()), CC_ZIP),
        getOrNull(ccAddress.getCountry(), CC_COUNTRY),
        getOrNull(ccAddress.getGmtOffset(), CC_GMT_OFFSET),
        getOrNull(ccTaxPercentage, CC_TAX_PERCENTAGE)
    };
  }
}

File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/scala/io/trino/tpcds/row/CallCenterRow.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.trino.tpcds.row

import java.util.{List => JList}

import io.trino.tpcds.`type`.{Address, Decimal}
import io.trino.tpcds.generator.CallCenterGeneratorColumn._
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.catalyst.expressions.GenericInternalRow

class CallCenterRow(
    private val ccCallCenterSk: Long,
    private val ccCallCenterId: String,
    private val ccRecStartDateId: Long,
    private val ccRecEndDateId: Long,
    private val ccClosedDateId: Long,
    private val ccOpenDateId: Long,
    private val ccName: String,
    private val ccClass: String,
    private val ccEmployees: Int,
    private val ccSqFt: Int,
    private val ccHours: String,
    private val ccManager: String,
    private val ccMarketId: Int,
    private val ccMarketClass: String,
    private val ccMarketDesc: String,
    private val ccMarketManager: String,
    private val ccDivisionId: Int,
    private val ccDivisionName: String,
    private val ccCompany: Int,
    private val ccCompanyName: String,
    private val ccAddress: Address,
    private val ccTaxPercentage: Decimal,
    nullBitMap: Long) extends TableRowWithNulls(nullBitMap, CC_CALL_CENTER_SK)
  with KyuubiTableRowWithNulls {

  override val nullBitMapInternal = nullBitMap
  override val firstColumnInternal = CC_CALL_CENTER_SK

  def getCcCallCenterSk: Long = ccCallCenterSk
  def getCcCallCenterId: String = ccCallCenterId
  def getCcRecStartDateId: Long = ccRecStartDateId
  def getCcRecEndDateId: Long = ccRecEndDateId
  def getCcClosedDateId: Long = ccClosedDateId
  def getCcOpenDateId: Long = ccOpenDateId
  def getCcName: String = ccName
  def getCcClass: String = ccClass
  def getCcEmployees: Int = ccEmployees
  def getCcSqFt: Int = ccSqFt
  def getCcHours: String = ccHours
  def getCcManager: String = ccManager
  def getCcMarketId: Int = ccMarketId
  def getCcMarketClass: String = ccMarketClass
  def getCcMarketDesc: String = ccMarketDesc
  def getCcMarketManager: String = ccMarketManager
  def getCcDivisionId: Int = ccDivisionId
  def getCcDivisionName: String = ccDivisionName
  def getCcCompany: Int = ccCompany
  def getCcCompanyName: String = ccCompanyName
  def getCcAddress: Address = ccAddress
  def getCcTaxPercentage: Decimal = ccTaxPercentage

  override def getValues: JList[String] = throw new UnsupportedOperationException()

  def internalRow: InternalRow =
    new GenericInternalRow(
      Array(
        getLongOrNull(ccCallCenterSk, CC_CALL_CENTER_SK),
        getStringOrNullInternal(ccCallCenterId, CC_CALL_CENTER_ID),
        getDateOrNullFromJulianDays(ccRecStartDateId, CC_REC_START_DATE_ID),
        getDateOrNullFromJulianDays(ccRecEndDateId, CC_REC_END_DATE_ID),
        getIntOrNullForKey(ccClosedDateId, CC_CLOSED_DATE_ID),
        getIntOrNullForKey(ccOpenDateId, CC_OPEN_DATE_ID),
        getStringOrNullInternal(ccName, CC_NAME),
        getStringOrNullInternal(ccClass, CC_CLASS),
        getIntOrNull(ccEmployees, CC_EMPLOYEES),
        getIntOrNull(ccSqFt, CC_SQ_FT),
        getStringOrNullInternal(ccHours, CC_HOURS),
        getStringOrNullInternal(ccManager, CC_MANAGER),
        getIntOrNull(ccMarketId, CC_MARKET_ID),
        getStringOrNullInternal(ccMarketClass, CC_MARKET_CLASS),
        getStringOrNullInternal(ccMarketDesc, CC_MARKET_DESC),
        getStringOrNullInternal(ccMarketManager, CC_MARKET_MANAGER),
        getIntOrNull(ccDivisionId, CC_DIVISION),
        getStringOrNullInternal(ccDivisionName, CC_DIVISION_NAME),
        getIntOrNull(ccCompany, CC_COMPANY),
        getStringOrNullInternal(ccCompanyName, CC_COMPANY_NAME),
        getStringOrNullInternal(ccAddress.getStreetNumber.toString, CC_STREET_NUMBER),
        getStringOrNullInternal(ccAddress.getStreetName, CC_STREET_NAME),
        getStringOrNullInternal(ccAddress.getStreetType, CC_STREET_TYPE),
        getStringOrNullInternal(ccAddress.getSuiteNumber, CC_SUITE_NUMBER),
        getStringOrNullInternal(ccAddress.getCity, CC_CITY),
        getStringOrNullInternal(ccAddress.getCounty, CC_ADDRESS),
        getStringOrNullInternal(ccAddress.getState, CC_STATE),
        getStringOrNullInternal(f"${ccAddress.getZip}%05d", CC_ZIP),
        getStringOrNullInternal(ccAddress.getCountry, CC_COUNTRY),
        getDecimalOrNull(ccAddress.getGmtOffset, CC_GMT_OFFSET, 5, 2),
        getDecimalOrNull(ccTaxPercentage, CC_TAX_PERCENTAGE, 5, 2)))
}

object CallCenterRow {
  class Builder {
    private var ccCallCenterSk = 0L
    private var ccCallCenterId: String = null
    private var ccRecStartDateId = 0L
    private var ccRecEndDateId = 0L
    private var ccClosedDateId = 0L
    private var ccOpenDateId = 0L
    private var ccName: String = null
    private var ccClass: String = null
    private var ccEmployees = 0
    private var ccSqFt = 0
    private var ccHours: String = null
    private var ccManager: String = null
    private var ccMarketId = 0
    private var ccMarketClass: String = null
    private var ccMarketDesc: String = null
    private var ccMarketManager: String = null
    private var ccDivisionId = 0
    private var ccDivisionName: String = null
    private var ccCompany = 0
    private var ccCompanyName: String = null
    private var ccAddress: Address = null
    private var ccTaxPercentage: Decimal = null
    private var nullBitMap = 0L

    def setCcCallCenterSk(ccCallCenterSk: Long): Builder = {
      this.ccCallCenterSk = ccCallCenterSk
      this
    }

    def setCcCallCenterId(ccCallCenterId: String): Builder = {
      this.ccCallCenterId = ccCallCenterId
      this
    }

    def setCcRecStartDateId(ccRecStartDateId: Long): Builder = {
      this.ccRecStartDateId = ccRecStartDateId
      this
    }

    def setCcRecEndDateId(ccRecEndDateId: Long): Builder = {
      this.ccRecEndDateId = ccRecEndDateId
      this
    }

    def setCcClosedDateId(ccClosedDateId: Long): Builder = {
      this.ccClosedDateId = ccClosedDateId
      this
    }

    def setCcOpenDateId(ccOpenDateId: Long): Builder = {
      this.ccOpenDateId = ccOpenDateId
      this
    }

    def setCcName(ccName: String): Builder = {
      this.ccName = ccName
      this
    }

    def setCcClass(ccClass: String): Builder = {
      this.ccClass = ccClass
      this
    }

    def setCcEmployees(ccEmployees: Int): Builder = {
      this.ccEmployees = ccEmployees
      this
    }

    def setCcSqFt(ccSqFt: Int): Builder = {
      this.ccSqFt = ccSqFt
      this
    }

    def setCcHours(ccHours: String): Builder = {
      this.ccHours = ccHours
      this
    }

    def setCcManager(ccManager: String): Builder = {
      this.ccManager = ccManager
      this
    }

    def setCcMarketId(ccMarketId: Int): Builder = {
      this.ccMarketId = ccMarketId
      this
    }

    def setCcMarketClass(ccMarketClass: String): Builder = {
      this.ccMarketClass = ccMarketClass
      this
    }

    def setCcMarketDesc(ccMarketDesc: String): Builder = {
      this.ccMarketDesc = ccMarketDesc
      this
    }

    def setCcMarketManager(ccMarketManager: String): Builder = {
      this.ccMarketManager = ccMarketManager
      this
    }

    def setCcDivisionId(ccDivisionId: Int): Builder = {
      this.ccDivisionId = ccDivisionId
      this
    }

    def setCcDivisionName(ccDivisionName: String): Builder = {
      this.ccDivisionName = ccDivisionName
      this
    }

    def setCcCompany(ccCompany: Int): Builder = {
      this.ccCompany = ccCompany
      this
    }

    def setCcCompanyName(ccCompanyName: String): Builder = {
      this.ccCompanyName = ccCompanyName
      this
    }

    def setCcAddress(ccAddress: Address): Builder = {
      this.ccAddress = ccAddress
      this
    }

    def setCcTaxPercentage(ccTaxPercentage: Decimal): Builder = {
      this.ccTaxPercentage = ccTaxPercentage
      this
    }

    def setNullBitMap(nullBitMap: Long): Unit = {
      this.nullBitMap = nullBitMap
    }

    def build(): CallCenterRow = {
      new CallCenterRow(
        ccCallCenterSk,
        ccCallCenterId,
        ccRecStartDateId,
        ccRecEndDateId,
        ccClosedDateId,
        ccOpenDateId,
        ccName,
        ccClass,
        ccEmployees,
        ccSqFt,
        ccHours,
        ccManager,
        ccMarketId,
        ccMarketClass,
        ccMarketDesc,
        ccMarketManager,
        ccDivisionId,
        ccDivisionName,
        ccCompany,
        ccCompanyName,
        ccAddress,
        ccTaxPercentage,
        nullBitMap)
    }
  }
}
